homeassistant/components/mqtt/light/schema_json.py

```diff
@@ -491,6 +509,43 @@ def _scale_rgbxx(self, rgbxx: tuple[int, ...], kwargs: Any) -> tuple[int, ...]:
             brightness = kwargs.get(ATTR_BRIGHTNESS, 255)
         return tuple(round(i / 255 * brightness) for i in rgbxx)
 
+    def _scale_brightness_to_device(self, brightness: int) -> int:
+        brightness_scale = self._config[CONF_BRIGHTNESS_SCALE]
+        brightness_scale_min = self._config[CONF_BRIGHTNESS_SCALE_MIN]
+        device_brightness: int = round(
+            (brightness - DEFAULT_BRIGHTNESS_SCALE_MIN)
+            / (DEFAULT_BRIGHTNESS_SCALE - DEFAULT_BRIGHTNESS_SCALE_MIN)
+            * (brightness_scale - brightness_scale_min)
+            + brightness_scale_min
+        )
+        # Make sure the brightness is not rounded down to 0
+        device_brightness = max(device_brightness, brightness_scale_min)
+        device_brightness = min(device_brightness, brightness_scale)
+
+        # Let's update mapping to in case multiple brightness values map to the same device value
+        if self._brightness_scale_map is None:
+            self._brightness_scale_map = generate_brightness_scale_map(
+                DEFAULT_BRIGHTNESS_SCALE_MIN,
+                DEFAULT_BRIGHTNESS_SCALE,
+                self._config[CONF_BRIGHTNESS_SCALE_MIN],
+                self._config[CONF_BRIGHTNESS_SCALE],
+            )
+
+            self._brightness_scale_map[device_brightness] = brightness
+
+        return device_brightness
+
+    def _scale_brightness_to_hass(self, device_brightness: int) -> int:
+        if self._brightness_scale_map is None:
+            self._brightness_scale_map = generate_brightness_scale_map(
+                DEFAULT_BRIGHTNESS_SCALE_MIN,
+                DEFAULT_BRIGHTNESS_SCALE,
+                self._config[CONF_BRIGHTNESS_SCALE_MIN],
+                self._config[CONF_BRIGHTNESS_SCALE],
+            )
+
+        return self._brightness_scale_map.get(device_brightness, 255)
+
```

Comment:
This is a lot of code for just scaling. All the current code lacks is that it might report back a value is 1/255 higher or lower as the value that would be sent.

Reply:
If you only talking about 254 scale, but if scale is 100, or 50, or 10 then all of it doesn't work

Reply:
May be sticking to a a scale factor (int) would be a better fit in case there are less then 255 steps? So in case of 254, the factor would be one 1 the offset would become 1.
If case we need to scale to 10 steps the scale factor could be round((255 - 5)/10) = 25 and the offset (calculated) would become 5 + 25 = 30 for step 1.
When calculating back brightness = value * scale factor + offset.
It would ensure the values are calculated correctly back and forward as we are using int's.
Only the first brightness step could have a different offset.

Reply:
It's advisable to conduct tests to verify the bidirectional functionality between A and B. Typically, the problems emerge when translating from B to A. This occurs if multiple A values correspond to a single B value, resulting in inconsistencies when reading back from the device. For instance, consider B = 50 mapping to A values of 127, 128, and 129. If you set the device to 127 but read back 129, it demonstrates this issue. To mitigate this, I implemented a dictionary to map these values and track the last sent request. In this example, it means storing 127 as equivalent to 50. While this approach may not be perfect, it should address most cases effectively.